docs/architecture/adr-005-subscription.md

```diff
@@ -0,0 +1,225 @@
+# ADR 5: Subscription
+
+## Changelog
+
+
+## Context
+
+Cosmos-SDK should allow developers to choose from a variety of payment models for their application. One of the most popular payment models for Internet applications, so it is natural to provide this payment model as a module for sdk users.
+
+For fuller context around this issue: see [\#4642](https://github.com/cosmos/cosmos-sdk/issues/4642)
+
+## Decision
+
+Create Subscription module in SDK so that users can subscribe to various subscription services. Subscription collectors will send messages to collect subscription payments from users who have due subscriptions. Subscriptions are paid out directly from a user account. If the user does not have enough funds to pay for the subscription is inactivated. Users can define maximum limits on how many periods the subscription is valid for; if this limit elapses before a subscription renewal, the subscription is invalidated.
+
+## Status
+
+#### Proposed
+
+## Implementation Changes
+
+Introduce the following messages to `x/subscription` module:
+
+```go
+// CreateSubscriptionMsg allows Service Provider to create new subscription service.
+type CreateSubscriptionMsg struct {
+    Name      string         // unique, human-readable name for subscription service
+    Amount    sdk.Coins      // amount to be collected for each subscription period
+    Period    time.time      // duration of subscription period
+    Collector sdk.AccAddress // address that will collect subscription payments
+}
+```
+
+```go
+// SubscribeMsg allows subscriber account to subscribe to provided service
+// If subscription to the service from user already exists, this msg is treated as renewal.
+// If msg.Limit != -1, then subscription.Limit += msg.Limit. Else, subscription.Limit = -1 (unlimited)
+type SubscribeMsg struct {
+    Name       string         // name of service to subscribe to
+    Subscriber sdk.AccAddress // address of subscriber
+    Limit      int64          // Maximum number of periods that subscription remains active. Limit = -1 implies no limit
```

Comment:
also, do we delete the subscription from the store once the limit is due? or do we move them to some other queue or smth and then we prune? 

I believe this requires a comment

Reply:
I was currently thinking just inactivate. If you look at the Collect handler, we pop the subscription off the queue if it is due (basically deleting from store). If the subscription is paid, it gets added back to end of queue.

If it is inactivated, the subscriptionID doesn't go back to end of queue, but it does stay in store as inactivated item.

Reply:
Ok so the reason I didn't delete is because I didn't want to traverse through the `DueQueue` and remove the deleted element. 

Though I think I can leave the subscriptionID in and just check if the retrieved subscription is nil. If it is, then I can remove it from the list